In [1]:
from __future__ import unicode_literals

import pandas as pd
import numpy as np
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

# 1. Preparing data - Latest 12 weeks of life of the banana plant
---

In [2]:
wind_speed_l12w = pd.read_csv('../../../../data/raw/FincaPorvenir/Metereologico/Latest_12-weeks_May-4_August-03_2018/' \
                                 'Velocidad-del-viento_May-4_August-03_2018.csv', )

In [3]:
print(wind_speed_l12w.shape)
wind_speed_l12w.head()

(4398, 2)


Fecha:  Velocidad del viento (Km/h)
0  2018-05-04 00:07:38                          0.0
1  2018-05-04 00:37:39                          0.0
2  2018-05-04 01:07:38                          0.0
3  2018-05-04 01:37:38                          0.0
4  2018-05-04 02:07:38                          0.0

## 1.1. We evaluate if this dataset has null type `NaN`
---

In [4]:
print(wind_speed_l12w.isnull().any())
wind_speed_l12w.isnull().values.any()

Fecha:                         False
Velocidad del viento (Km/h)    False
dtype: bool


False

Don't have null values

## 1.2. Selecting  relevant index columns features
---

Since the dataset has a column called **`Fecha:`**, which is not a numerical value,

it will be removed so that it does not interfere **with our subsequent scaling**, 

so we are only going to reference the values or samples of the column

** `Velocidad del viento (Km/h)` ** and assigning them to the matrix 

`wind_speed_l12w_array` created such as follow:

In [5]:
wind_speed_l12w_array = wind_speed_l12w.iloc[:,1].values

In [6]:
# Rehape the luminosity_luxes array 
wind_speed_l12w_array = wind_speed_l12w_array.reshape(-1, 1)
wind_speed_l12w_array

array([[0.        ],
       [0.        ],
       [0.        ],
       ...,
       [0.        ],
       [0.        ],
       [4.21052632]])

## 1.3 Generating descriptive data to dataset
---

In [7]:
col = ['Velocidad del viento (Km/h)']
wind_speed_l12w_values_df = pd.DataFrame(wind_speed_l12w_array, columns=col)
wind_speed_l12w_values_df_describe = wind_speed_l12w_values_df.describe()

In [8]:
wind_speed_l12w_values_df_describe

Velocidad del viento (Km/h)
count                  4398.000000
mean                      1.297827
std                       2.892122
min                       0.000000
25%                       0.000000
50%                       0.000000
75%                       1.814747
max                      68.571429

In [9]:
# Export this descriptive data to comma separated values and java script object notation
wind_speed_l12w_values_df_describe.to_csv('../../../../data/interim/WindSpeed/Latest_12-weeks_May-4_August-03_2018/' +'\n' 
                              'Wind-Speed_Describe_May-4_August-03_2018.cvs', sep=',', header=True, index=True)
wind_speed_l12w_values_df_describe.to_json('../../../../data/interim/WindSpeed/Latest_12-weeks_May-4_August-03_2018/' +'\n'
                                     'Wind-Speed_Describe_May-4_August-03_2018.json')

# 2. Creating Temperature Training and Testing datasets

We have a **`wind_speed_l12w_values_df`** dataset with 4398 samples rows.

In [10]:
print(wind_speed_l12w_values_df.shape)
wind_speed_l12w_values_df.head()

(4398, 1)


Velocidad del viento (Km/h)
0                          0.0
1                          0.0
2                          0.0
3                          0.0
4                          0.0

We'll divide it into two differents datasets:

- Training dataset
- Testing dataset

This is executed through [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html  "sklearn.model_selection.train_test_split") function of this way:

`train_test_split` receive as a data parameter a numpy array, we have to turn the 

**`wind_speed_l12w_values_df`** dataframe to numpy array such as follow:

In [11]:
# numpy_wind_speed_l12w_values_df = wind_speed_l12w_values_df.reset_index().values
numpy_wind_speed_l12w_values_df = wind_speed_l12w_values_df.values

In [12]:
numpy_wind_speed_l12w_values_df

array([[0.        ],
       [0.        ],
       [0.        ],
       ...,
       [0.        ],
       [0.        ],
       [4.21052632]])

We compose the following datasets from **`numpy_wind_speed_l12w_values_df`** array :

- `wind_speed_l12w_values_train`, which is the training matrix
- `wind_speed_l12w_values_test`, the testing matrix

We use the [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html  "sklearn.model_selection.train_test_split") function to create the training and testing dataset. Their parameters are:

![alt text](https://cldup.com/hukPWvvLxt-3000x3000.png "klearn.model_selection.train_test_split")


- The first parameter **should be an array**, then we pass the **`numpy_wind_speed_l12w_values_df`** which contain all column features of wind speed.

Of this way, we pass all data (4398 samples rows) to from them `wind_speed_l12w_values_train`

and `wind_speed_l12w_values_test` will be created

- The `test_size=0.5` parameter means a 50% division; which means 
that half of the data goes to the test dataset and the other half 
goes to the training dataset.

A good choice for the size of tests **is usually 0.2 ie 20% or 0.25 or even 30%.** 

In some rare cases we will have 40% but almost never 0.5 or 50%

We choose 20% which means that we will have 20% of 9752 samples or observations for the test data set, 

in this case **4390 * 0.2 = 879.6 samples or records for the test data set**


- The `train_size` parameter is the training dataset size.  **`test_size + train_size = 1 or 100%`**, 
then isn't necessary include it, because wheter we include to `test_size = 0.2`, then the remaining 
data will be to `train_size` this means **0.8 or 80%**

This means that **4390 * 0.8 = 3518.4  training dataset samples rows. **


- random_state is a seed or data source for generating random values for the data sets. 

If this parameter is not passed, the data will be generated in a random way, but in the way as by default numpy works them.

### 2.1  Creating Training and testing wind speed datasets
---

- `wind_speed_l12w_values_train`, which is the training matrix
- `wind_speed_l12w_values_test`, the testing matrix

In [13]:
wind_speed_l12w_values_train, wind_speed_l12w_values_test = train_test_split(numpy_wind_speed_l12w_values_df, test_size = 0.2)

In [14]:
# We have 3518,4 rows to luminosity_luxes_train
print(type(wind_speed_l12w_values_train))
print("The dimensionality of wind speed training dataset is: " +'\n' , wind_speed_l12w_values_train.shape)
print('\n')

# And we have 880 rows to luminosity_luxes_test
print(type(wind_speed_l12w_values_test))
print("The dimensionality of wind speed testing dataset is: " +'\n' , wind_speed_l12w_values_test.shape)

<class 'numpy.ndarray'>
The dimensionality of wind speed training dataset is: 
 (3518, 1)


<class 'numpy.ndarray'>
The dimensionality of wind speed testing dataset is: 
 (880, 1)


In this way, the model, as you progressively learn the correlations in the training set, the better the prediction of the results in the test set.

But if the model learns too much from memory the correlations of the training sets,

that is to say, when one learns from memory and does not understand things, then it

will have problems to predict what is happening on the set of tests, because it is

learned for difficult correlations, if the logic is not well understood and you can

not make good predictions. This is called overfitting or overfitting

The really important thing is to understand that we need to have two different datasets

- Training set with which the ML model learns
- Test set, on which we test whether the ML model correctly learned the correlations


---
##  3. Feature Scaling Temperature training and testing dataset  
---

[This article post](http://benalexkeen.com/feature-scaling-with-scikit-learn/  "Feature Scaling with scikit-learn") it's a great reference to explore the features scaling methods
on scikit learn

- `StandardScaler` assume that data is normally distributed at the level of each characteristic or variable. If the data is not normally distributed, it is not the best alternative to use for scaling.  

- `Min-Max Scaler` it is probably the most famous scaling algorithm and what it does is resize the range to leave it in a dimension of 0 to 1 or -1 and 1 (in case there are negative values in the original dataset of input data)

This scale of maximums and minimums works best for cases where standard scaling may not work properly. If the distribution is not Gaussian or the standard deviation is very small the escalation of maximums and minimums is the best idea.
However, it is sensitive to outliers or outliers, so if there are outliers in the data it is better to consider robust scaling.

- `Robust Scaler` it is similar to the previous one of maximums and minimums, only that it uses interquartile ranges instead of maximums and minimums, which makes it robust for the outliers

- `Normalizer` which scales each value, dividing each value by its magnitude in n dimensional spaces for n number of characteristics.


---

We will use the scaling of maximums and minimums to scale the water table data, because the standard deviation is very small, 

it does not have atypical values and it does not follow a normal distribution (you have to check this)

We apply the maximum and minimum scaling. We provide a rank or base scale that **will be between 0 and 1** using an object,

[MinMaxScaler](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html "sklearn.preprocessing.MinMaxScaler") which transforms each characteristic, (in this 

case it will be the columns feature of `Velocidad del viento (Km/h)`) individually according to a given range.

Product of its applicability, generates these attributes in the dataset, already transformed: 

![alt text](https://cldup.com/lTIv4HXgTk-3000x3000.png "sklearn.preprocessing.MinMaxScaler")

### 3.1 We apply maximium and minimum feature scaling to Wind speed training dataset

In [15]:
# We provide a base scale range
scaler = MinMaxScaler(feature_range=(0, 1))

print("Remember our temperature training data " + '\n', wind_speed_l12w_values_train)

Remember our temperature training data 
 [[0.        ]
 [1.08679245]
 [1.94070081]
 ...
 [1.64759725]
 [0.        ]
 [0.        ]]


With the [fit](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler.fit  "MinMaxScaler.fit")
we compute the maximum and minimum value of  `wind_speed_l12w_values_train` dataset to be used in the subsequent scaling 

We assing these values to `minmax_scale_training` variable.

In [16]:
minmax_scale_training = scaler.fit(wind_speed_l12w_values_train.astype(float))
# print(minmax_scale_training.data_max_)
# http://terrapinssky.blogspot.com/2017/10/pythonresolved-dataconversionwarning.html

Then, we apply the [transform](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler.transform "MinMaxScaler.transform") method to transform these data to maximum and mínimum scale value. 

Here, with this process, the `wind_speed_l12w_values_train` data are scaled between **0 to 1**  selected range

In [17]:
# transform precipitations_train data to maximum and mínimum scale value. 
wind_speed_l12w_minmax_training = minmax_scale_training.transform(wind_speed_l12w_values_train)

In [18]:
print ("And, these are our scaled data: " + '\n')
wind_speed_l12w_minmax_training

And, these are our scaled data: 



array([[0.        ],
       [0.02188679],
       [0.03908356],
       ...,
       [0.03318078],
       [0.        ],
       [0.        ]])

In [19]:
print('Wind speed Training dataset. Minimum value after MaxMinScaler:\nWind speed={:.1f}'
      .format(wind_speed_l12w_minmax_training[:,0].min()))

print('Wind speed Training dataset. Maximum value after MaxMinScaler:\nWind speed={:.1f}'
      .format(wind_speed_l12w_minmax_training[:,0].max()))

Wind speed Training dataset. Minimum value after MaxMinScaler:
Wind speed=0.0
Wind speed Training dataset. Maximum value after MaxMinScaler:
Wind speed=1.0


Then, our  MinMaxScaler normalized training dataset is `wind_speed_l12w_minmax_training` numpy array

- We export this array to comma separated values 

In [20]:
wind_speed_l12w_train_df = pd.DataFrame(wind_speed_l12w_minmax_training, columns=col)

In [21]:
print(wind_speed_l12w_train_df.shape)
wind_speed_l12w_train_df.head()

(3518, 1)


Velocidad del viento (Km/h)
0                     0.000000
1                     0.021887
2                     0.039084
3                     0.000000
4                     0.000000

In this way we have the dataset **`wind_speed_l12w_train_df`** standardized and training, and export it to a .csv file

In [22]:
wind_speed_l12w_train_df.to_csv('../../../../data/processed/WindSpeed/Latest_12-weeks_May-4_August-03_2018/' +'\n' 
                                 'Wind-Speed_Normalized_TRAINING_May-4_August-03_2018.csv', sep=',', header=True, index=False)

### 3.2 We apply maximium and minimum feature scaling to Temperature testing dataset


In [23]:
# We provide a base scale range
scaler = MinMaxScaler(feature_range=(0, 1))

print("Remember our temperature testing data " + '\n', wind_speed_l12w_values_test)

Remember our temperature testing data 
 [[ 8.32369942]
 [ 1.90728477]
 [ 1.19800333]
 [ 1.15569823]
 [ 0.        ]
 [ 4.3373494 ]
 [ 0.        ]
 [ 1.18421053]
 [ 2.30031949]
 [ 0.        ]
 [ 2.19512195]
 [ 2.40400668]
 [ 1.86770428]
 [ 0.        ]
 [ 0.        ]
 [ 1.29147982]
 [ 1.39670223]
 [ 0.        ]
 [ 0.        ]
 [ 2.30769231]
 [ 0.        ]
 [ 1.61979753]
 [ 1.62711864]
 [ 2.07492795]
 [ 2.34527687]
 [ 0.        ]
 [ 0.        ]
 [ 0.        ]
 [ 0.        ]
 [ 3.24324324]
 [ 0.        ]
 [ 1.55339806]
 [ 0.        ]
 [ 0.        ]
 [ 0.        ]
 [ 0.        ]
 [ 1.48914167]
 [ 0.        ]
 [ 0.        ]
 [ 0.        ]
 [ 0.        ]
 [ 3.67346939]
 [ 1.57894737]
 [ 0.        ]
 [ 2.89156626]
 [ 0.        ]
 [ 1.2642669 ]
 [ 0.        ]
 [ 1.10260337]
 [ 0.        ]
 [ 0.        ]
 [ 0.        ]
 [ 1.38594803]
 [ 1.22657581]
 [ 1.56521739]
 [ 1.43712575]
 [ 1.4968815 ]
 [ 1.86770428]
 [ 1.9862069 ]
 [ 0.        ]
 [ 1.96452933]
 [ 2.1719457 ]
 [ 0.        ]
 [ 0.        ]


In [24]:
minmax_scale_test = scaler.fit(wind_speed_l12w_values_test.astype(float))
# transform luminosity_luxes_test data to maximum and mínimum scale value. 
wind_speed_l12w_minmax_test = minmax_scale_test.transform(wind_speed_l12w_values_test)

In [25]:
print ("And, these are our testing scaled data: " + '\n')
wind_speed_l12w_minmax_test

And, these are our testing scaled data: 



array([[0.12138728],
       [0.02781457],
       [0.01747088],
       [0.01685393],
       [0.        ],
       [0.06325301],
       [0.        ],
       [0.01726974],
       [0.03354633],
       [0.        ],
       [0.0320122 ],
       [0.03505843],
       [0.02723735],
       [0.        ],
       [0.        ],
       [0.01883408],
       [0.02036857],
       [0.        ],
       [0.        ],
       [0.03365385],
       [0.        ],
       [0.02362205],
       [0.02372881],
       [0.03025937],
       [0.03420195],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.0472973 ],
       [0.        ],
       [0.02265372],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.02171665],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.05357143],
       [0.02302632],
       [0.        ],
       [0.04216867],
       [0.        ],
       [0.01843723],
       [0.   

In [26]:
print('Wind Speed Testing dataset. Minimum value after MaxMinScaler:\nWind Speed={:.1f}'
      .format(wind_speed_l12w_minmax_test[:,0].min()))

print('Wind Speed Testing dataset. Maximum value after MaxMinScaler:\nWind Speed={:.1f}'
      .format(wind_speed_l12w_minmax_test[:,0].max()))

Wind Speed Testing dataset. Minimum value after MaxMinScaler:
Wind Speed=0.0
Wind Speed Testing dataset. Maximum value after MaxMinScaler:
Wind Speed=1.0


Then, our  MinMaxScaler normalized training dataset is `wind_speed_l12w_minmax_test` numpy array

- We export this array to comma separated values

In [27]:
wind_speed_l12w_test_df = pd.DataFrame(wind_speed_l12w_minmax_test, columns=col)

In [28]:
print(wind_speed_l12w_test_df.shape)
wind_speed_l12w_test_df

(880, 1)


Velocidad del viento (Km/h)
0                       0.121387
1                       0.027815
2                       0.017471
3                       0.016854
4                       0.000000
5                       0.063253
6                       0.000000
7                       0.017270
8                       0.033546
9                       0.000000
10                      0.032012
11                      0.035058
12                      0.027237
13                      0.000000
14                      0.000000
15                      0.018834
16                      0.020369
17                      0.000000
18                      0.000000
19                      0.033654
20                      0.000000
21                      0.023622
22                      0.023729
23                      0.030259
24                      0.034202
25                      0.000000
26                      0.000000
27                      0.000000
28                      0.000000
29                      0.047297
..                           ...
850                     0.016562
851                     0.048499
852                     0.000000
853                     0.047727
854                     0.015719
855                     0.000000
856                     0.000000
857                     0.000000
858                     0.034036
859                     0.000000
860                     0.000000
861                     0.000000
862                     0.000000
863                     0.000000
864                     0.000000
865                     0.039326
866                     0.109375
867                     0.000000
868                     0.036207
869                     0.000000
870                     0.000000
871                     0.091703
872                     0.000000
873                     0.000000
874                     0.000000
875                     0.000000
876                     0.026515
877                     0.000000
878                     0.029494
879                     0.020690

[880 rows x 1 columns]

In [29]:
wind_speed_l12w_test_df.to_csv('../../../../data/processed/WindSpeed/Latest_12-weeks_May-4_August-03_2018/' +'\n' 
                                 'Wind-Speed_Normalized_TESTING_May-4_August-03_2018.csv', sep=',', header=True, index=False)